zrml/asset-router/src/lib.rs

```diff
@@ -0,0 +1,742 @@
+            } else if let Ok(asset) = T::CampaignAssetType::try_from(currency_id) {
+                T::CampaignAssets::reducible_balance(asset, who, false) >= value
+            } else if let Ok(asset) = T::CustomAssetType::try_from(currency_id) {
+                T::CustomAssets::reducible_balance(asset, who, false) >= value
+            } else {
+                false
+            }
+        }
+
+        fn slash(
+            currency_id: Self::CurrencyId,
+            who: &T::AccountId,
+            amount: Self::Balance,
+        ) -> Self::Balance {
+            if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
+                <T::Currencies as MultiCurrency<T::AccountId>>::slash(currency, who, amount)
+            } else if let Ok(asset) = T::MarketAssetType::try_from(currency_id) {
+                T::MarketAssets::slash(asset, who, amount)
+                    .map(|b| amount.saturating_sub(b))
+                    .unwrap_or_else(|_| amount)
+            } else if let Ok(asset) = T::CampaignAssetType::try_from(currency_id) {
+                T::CampaignAssets::slash(asset, who, amount)
+                    .map(|b| amount.saturating_sub(b))
+                    .unwrap_or_else(|_| amount)
+            } else if let Ok(asset) = T::CustomAssetType::try_from(currency_id) {
+                T::CustomAssets::slash(asset, who, amount)
+                    .map(|b| amount.saturating_sub(b))
+                    .unwrap_or_else(|_| amount)
+            } else {
+                amount
+            }
+        }
+    }
+
+    impl<T: Config> MultiCurrencyExtended<T::AccountId> for Pallet<T> {
+        type Amount = <T::Currencies as MultiCurrencyExtended<T::AccountId>>::Amount;
+
+        fn update_balance(
+            currency_id: Self::CurrencyId,
+            who: &T::AccountId,
+            by_amount: Self::Amount,
+        ) -> DispatchResult {
+            if let Ok(currency) = T::CurrencyType::try_from(currency_id) {
+                return <T::Currencies as MultiCurrencyExtended<T::AccountId>>::update_balance(
+                    currency, who, by_amount,
+                );
+            }
+
+            if by_amount.is_zero() {
+                return Ok(());
+            }
+
+            // Ensure this doesn't overflow. There isn't any traits that exposes
+            // `saturating_abs` so we need to do it manually.
+            let by_amount_abs = if by_amount == Self::Amount::min_value() {
+                Self::Amount::max_value()
+            } else {
+                by_amount.abs()
+            };
+
+            let by_balance = TryInto::<Self::Balance>::try_into(by_amount_abs)
+                .map_err(|_| Error::<T>::AmountIntoBalanceFailed)?;
+            if by_amount.is_positive() {
+                Self::deposit(currency_id, who, by_balance)
+            } else {
+                Self::withdraw(currency_id, who, by_balance).map(|_| ())
+            }
```

Comment:
I don't fully get this.

Reply:
the `Amount` type is [`Signed`](https://github.com/zeitgeistpm/open-runtime-module-library/blob/master/currencies/src/lib.rs#L686), but the non currency asset implementation only provide updating balances via the `Mutate` trait, which requires a signed number.

The most common signed number representation utilizes the [Two's complement](https://en.wikipedia.org/wiki/Two%27s_complement). Effectively, more negative numbers can be represented and positive numbers, to be more precise `max_neg = -max_pos - 1`. To avoid overflows, i.e. converting a negative number using `abs()` into a positive number that is out of the range of possible positive numbers of the `Signed` type, a case distinction is used. Whenever `by_amount` is `max_neg`, `Amount::max_value()` is used instead, which results in `max_pos` (`abs(max_neg) - 1`).